### DISCUSSION TITLE
What is the recommend way to access db via hooks

### DISCUSSION THREAD

**Discussion by DeanKamali** [directus/directus#6409] [Archive]
I wonder if there is a recommended way to update the database from within a hook after a change is made to the collection. 

For example, I have an `invoice` collection, and after inserting a new item, I would like to update some properties for that item, things like total balance, remaining balance ..etc  

I have been trying to import the SDK in `items.create.after ` hook, but keep getting 

```
16:06:36 ⚠️  Couldn't register hook "updates."
16:06:36 ⚠️  Cannot use import statement outside a module
```

**Comment by philleepflorence** [comment]
Hi DeanKamali, How did you structure your hook compared to the [Custom API Hooks Guide: Directus Docs](https://docs.directus.io/guides/api-hooks/#_4-develop-your-custom-hook) and what D9 version are you using?
In the meantime, you could try to: 
- access the database using the `services` parameter `const { ItemsService } = services;` or 
- update the payload data using the `before` event.

**Reply by DeanKamali** [reply]
Thanks @philleepflorence, I’m able to intercept the payload and make required changes via `before` returning it. 

A side question, what if I wanted to update a different collection after a change is made to a certain collection? 

Should I use the sdk, API? Or there is a way to access other collections from within a hook? 

I’m using the latest version of directus.

**Reply by rijkvanzanten** [reply]
> Should I use the sdk, API? Or there is a way to access other collections from within a hook?

The custom endpoints docs actually has a good example on how to read data from another collection. The same logic applies to a hook:

https://docs.directus.io/guides/api-endpoints/#full-example

**Reply by AlexVanK** [reply]
@DeanKamali (@rijkvanzanten - https://github.com/directus/directus/discussions/10067)
were you able to figure this out?

I am trying to read (and update) another collection information, from the one that is firing the hook...

I don't understand how the same logic for a custom endpoint applies also to custom hooks...
![image](https://user-images.githubusercontent.com/56647386/143617722-b475fc78-c207-4856-9409-4523e1559617.png)

But the `registerHook()` function doesn't have the `router` parameter, so I cannot use that to read data.

Does this mean that:

- I have to create a custom endpoint? (I don't really need that, I could use axios and query normally with SDK)... or
- I have to create an endpoint withing the hook file? ... or
- Is there another way to access other data from within a hook?

---

Also, once I am able to read another collection's data, how do I go about updating it?

**Reply by rijkvanzanten** [reply]
> But the registerHook() function doesn't have the router parameter, so I cannot use that to read data.

You don't have to. The router object is only required to register the actual endpoints. Everything else uses that context parameter, that's the same for hooks and endpoints.

> I have to create a custom endpoint? (I don't really need that, I could use axios and query normally with SDK)... or

Nope

> I have to create an endpoint withing the hook file? ... or

Nope

> Is there another way to access other data from within a hook?

Nope, the way to access data is the mostly the same from how you'd do it in a custom endpoint. You can use the services provided in the context object.

The [hooks full example](https://docs.directus.io/extensions/hooks/#example-sync-with-external) shows how to use a service, and the [endpoints full example](https://docs.directus.io/extensions/endpoints/#example-recipes) shows how you can use the `ItemsService` to read an item.

When you combine the two, you'll end up with something like:

```js
module.exports = registerHook({ action }, { services }) {
  const { ItemsService } = services;

  action('items.read', async ({ schema }) => {
    const articles = new ItemsService('articles', { schema });

    // Read an item
    const item = await articles.readOne(15);

    // Update an item
    await articles.updateOne(15, { status: 'published' });
  });
}
```

**Comment by mhdaouas** [comment]
How to update the item? Say I want to update the status value in a workflow based on a hook. Should I do it in the "before" or "after" hook and with which method/object can I update? It is not clear at all in the actual doc.

**Reply by mhdaouas** [reply]
OK never mind, you just need to do this:
`return Object.assign(item, { <key>: <val> });`

like described here: https://github.com/directus/directus/issues/3776
